pkg/controller/networkpolicy/endpoint_querier.go

```diff
@@ -18,81 +18,99 @@
+	// along with the list of NetworkPolicy ingress/egress rules which select the provided Pod.
+	QueryNetworkPolicyRules(namespace, podName string) (*antreatypes.EndpointNetworkPolicyRules, error)
 }
 
-// endpointQuerier implements the EndpointQuerier interface
-type endpointQuerier struct {
+// EndpointQuerierImpl implements the EndpointQuerier interface
+type EndpointQuerierImpl struct {
 	networkPolicyController *NetworkPolicyController
 }
 
-// EndpointQueryResponse is the reply struct for anctl endpoint queries
-type EndpointQueryResponse struct {
-	Endpoints []Endpoint `json:"endpoints,omitempty"`
+// NewEndpointQuerier returns a new *EndpointQuerierImpl.
+func NewEndpointQuerier(networkPolicyController *NetworkPolicyController) *EndpointQuerierImpl {
+	return &EndpointQuerierImpl{
+		networkPolicyController: networkPolicyController,
+	}
 }
 
-type Endpoint struct {
-	Namespace string   `json:"namespace,omitempty"`
-	Name      string   `json:"name,omitempty"`
-	Policies  []Policy `json:"policies,omitempty"`
-	Rules     []Rule   `json:"rules,omitempty"`
+// PolicyRuleQuerier handles requests for querying effective policy rule on entities.
+type PolicyRuleQuerier interface {
+	QueryNetworkPolicyEvaluation(entities *controlplane.NetworkPolicyEvaluationRequest) (*controlplane.NetworkPolicyEvaluationResponse, error)
 }
 
-type PolicyRef struct {
-	Namespace string    `json:"namespace,omitempty"`
-	Name      string    `json:"name,omitempty"`
-	UID       types.UID `json:"uid,omitempty"`
+// policyRuleQuerier implements the PolicyRuleQuerier interface
+type policyRuleQuerier struct {
+	endpointQuerier EndpointQuerier
 }
 
-type Policy struct {
-	PolicyRef
+// NewPolicyRuleQuerier returns a new *policyRuleQuerier
+func NewPolicyRuleQuerier(endpointQuerier EndpointQuerier) *policyRuleQuerier {
+	return &policyRuleQuerier{
+		endpointQuerier: endpointQuerier,
+	}
 }
 
-type Rule struct {
-	PolicyRef
-	Direction cpv1beta.Direction `json:"direction,omitempty"`
-	RuleIndex int                `json:"ruleindex,omitempty"`
+type lessFunc func(p1, p2 *antreatypes.RuleInfo) int
+
+// ByRulePriority implements the Sort interface, sorting the rules within.
+// Comparators should be ordered by their importance in terms of determining rule priority.
+type ByRulePriority struct {
+	rules       []*antreatypes.RuleInfo
+	comparators []lessFunc
 }
 
-// NewEndpointQuerier returns a new *endpointQuerier.
-func NewEndpointQuerier(networkPolicyController *NetworkPolicyController) *endpointQuerier {
-	n := &endpointQuerier{
-		networkPolicyController: networkPolicyController,
+func (s ByRulePriority) Len() int { return len(s.rules) }
+
+func (s ByRulePriority) Swap(i, j int) { s.rules[i], s.rules[j] = s.rules[j], s.rules[i] }
+
+func (s ByRulePriority) Less(i, j int) bool {
+	p, q := s.rules[i], s.rules[j]
+	// Try all but the last comparison.
+	var k int
+	for k = 0; k < len(s.comparators)-1; k++ {
```

Comment:
can't it use `len(s.comparators)` then return `false` if the loop doesn't determine a result?